pkg/controller/endpointslicemirroring/endpointslicemirroring_controller.go

```diff
@@ -0,0 +1,471 @@
+	// synced at least once. Added as a member to the struct to allow injection
+	// for testing.
+	endpointsSynced cache.InformerSynced
+
+	// endpointSliceLister is able to list/get endpoint slices and is populated
+	// by the shared informer passed to NewController
+	endpointSliceLister discoverylisters.EndpointSliceLister
+	// endpointSlicesSynced returns true if the endpoint slice shared informer
+	// has been synced at least once. Added as a member to the struct to allow
+	// injection for testing.
+	endpointSlicesSynced cache.InformerSynced
+	// endpointSliceTracker tracks the list of EndpointSlices and associated
+	// resource versions expected for each Endpoints resource. It can help
+	// determine if a cached EndpointSlice is out of date.
+	endpointSliceTracker *endpointSliceTracker
+
+	// serviceLister is able to list/get services and is populated by the shared
+	// informer passed to NewController.
+	serviceLister corelisters.ServiceLister
+	// servicesSynced returns true if the services shared informer has been
+	// synced at least once. Added as a member to the struct to allow injection
+	// for testing.
+	servicesSynced cache.InformerSynced
+
+	// reconciler is an util used to reconcile EndpointSlice changes.
+	reconciler *reconciler
+
+	// Endpoints that need to be updated. A channel is inappropriate here,
+	// because it allows Endpoints with lots of addresses to be serviced much
+	// more often than Endpoints with few addresses; it also would cause an
+	// Endpoints resource that's inserted multiple times to be processed more
+	// than necessary.
+	queue workqueue.RateLimitingInterface
+
+	// maxEndpointsPerSlice references the maximum number of endpoints that
+	// should be added to an EndpointSlice
+	maxEndpointsPerSlice int32
+
+	// workerLoopPeriod is the time between worker runs. The workers process the
+	// queue of changes to Endpoints resources.
+	workerLoopPeriod time.Duration
+
+	// endpointUpdatesBatchPeriod is an artificial delay added to all Endpoints
+	// syncs triggered by EndpointSlice changes. This can be used to reduce
+	// overall number of all EndpointSlice updates.
+	endpointUpdatesBatchPeriod time.Duration
+}
+
+// Run will not return until stopCh is closed.
+func (c *Controller) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	klog.Infof("Starting EndpointSliceMirroring controller")
+	defer klog.Infof("Shutting down EndpointSliceMirroring controller")
+
+	if !cache.WaitForNamedCacheSync("endpoint_slice_mirroring", stopCh, c.endpointsSynced, c.endpointSlicesSynced, c.servicesSynced) {
+		return
+	}
+
+	klog.V(2).Infof("Waiting for application of %s label", discovery.LabelSkipMirror)
+	c.waitForSkipLabelApplication()
+	klog.V(2).Infof("%s label successfully applied", discovery.LabelSkipMirror)
+
+	for i := 0; i < workers; i++ {
+		go wait.Until(c.worker, c.workerLoopPeriod, stopCh)
+	}
+
+	go func() {
```

Comment:
Whats the point in spinning of a separate goroutine which just does a deferred handle crash, especially as this run method already has a deferred handle crash.

Reply:
🤦 that was apparently a relic from the Endpoints controller - there was actually a function call in the goroutine. Somehow I managed to remove the function call but left the pointless goroutine. Thanks for catching this one!